Fixes Container, Event Manager, CLI & Rewrite rules for AVIF

remyperona opened the pull request:
## Description

In this PR:
- Start usage of the container & event manager
- Fix issue with the CLI command for next gen images generation
- Fix issue with rewrite rules for WebP & AVIF
- Fix the picture display and change it to a subscriber
- Fix some function names

## Type of change
- Bug fix (non-breaking change which fixes an issue).
# Checklists

## Generic development checklist

- [x] My code follows the style guidelines of this project, with adapted comments and without new warnings.
- [ ] I have added unit and integration tests that prove my fix is effective or that my feature works.
- [ ] The CI passes locally with my changes (including unit tests, integration tests, linter).

## Test summary

- [x] I triggered all changed lines of code at least once without new errors/warnings/notices.

remyperona committed:
fix next gen CLI

remyperona committed:
update options name

remyperona committed:
update options name

remyperona committed:
update option name

remyperona committed:
fix function name

remyperona committed:
fix function name

remyperona committed:
fix options name usage

remyperona committed:
update class name

remyperona committed:
add container loading

remyperona committed:
add service providers and subscribers loading

remyperona committed:
move picture display instantiation

remyperona committed:
Merge branch 'feature/avif' into task/avif-fixes

remyperona committed:
fix return hint

remyperona committed:
update naming

remyperona committed:
move webp to service provider

remyperona committed:
cleanup classes

remyperona committed:
add service provider

remyperona committed:
convert classes to subscribers

remyperona committed:
re-add instance getter

remyperona committed:
move CDN methods to CDN class

remyperona committed:
add filter

remyperona committed:
fix condition

remyperona committed:
remove unused class

remyperona committed:
add AVIF service provider

remyperona committed:
update option check

remyperona committed:
remove webp from extensions list

remyperona committed:
remove avif from extensions list

remyperona committed:
add missing classes

MathieuLamiot commented:
@Miraeld or @Khadreal to check this one to unlock the tests

remyperona committed:
update returned values depending on avif option

CrochetFeve0251 commented on classes/Plugin.php:
```diff
@@ -41,6 +81,24 @@ public function __construct( $plugin_args ) {
+
+		$this->container->add(
+			'event_manager',
+			function () {
+				return new EventManager();
+			}
```
Why are we using a method here instead of using the classname directly?
Also is it possible to share it?

CrochetFeve0251 commented on inc/classes/Dependencies/wp-media/event-manager/event-manager-aware-subscriber-interface.php:
```diff
@@ -1,14 +1,14 @@
 <?php
```
Can we change the name of the files to respect PSR4?

CrochetFeve0251 commented on classes/Plugin.php:
```diff
@@ -171,4 +237,39 @@ public function maybe_activate() {
+			'Imagify\CDN\ServiceProvider',
+			'Imagify\Picture\ServiceProvider',
+			'Imagify\Webp\ServiceProvider',
+		];
```
Can we externalize that out from the plugin class?

CrochetFeve0251 commented on classes/Plugin.php:
```diff
@@ -171,4 +237,39 @@ public function maybe_activate() {
+	 * @return void
+	 */
+	private function load_subscribers( ServiceProviderInterface $service_provider ) {
+		if ( empty( $service_provider->subscribers ) ) {
```
Is it possible to use a method instead of the property `subscribers` so it will be possible to create an interface later?

remyperona committed:
rename files

remyperona committed:
add get_subscribers() method

remyperona committed:
share event manager instance and use new get_subscribers() method

remyperona committed:
move service providers list to config file

Mai-Saad commented:
@Tabrisrp Thanks for the PR. Please find exploratory notes below: (tested on nginx so far)
1- After a fresh install, if we enable Display Next-Gen -> rewrite rules, we will have this warning in debug.log and admin will have this (note: rewrite rules added in conf)
`[09-Feb-2024 05:35:49 UTC] PHP Warning:  Undefined variable $result in /var/www/imagify.rocketlabsqa.ovh/htdocs/wp-content/plugins/imagify-plugin/classes/Avif/RewriteRules/Display.php on line 79`
![Screenshot from 2024-02-09 07-37-09](https://github.com/wp-media/imagify-plugin/assets/76941962/91bc45dd-9926-472f-85d7-7e007242d669)
2- Whenever we change anything in settings and save, we will see the bulk optimization notice. Is this expected? /cc @piotrbak (this is not on trunk)
![Screenshot from 2024-02-09 07-37-43](https://github.com/wp-media/imagify-plugin/assets/76941962/5417ee6e-59a1-4833-95a2-419c4da29906)
3- Rewrite rules added to conf is as below, shouldn't we replace `webp` by `next-gen` in the 1st and last line? Shouldn't the location line have webp along with avif? Not sure about the lines in between, @piotrbak what do you think? Is there a sample of expected rules?
```
# BEGIN Imagify: rewrite rules for webp
location ~* ^(/.+)\.(jpg|jpeg|jpe|png|gif|avif)$ {
	add_header Vary Accept;

	if ($http_accept ~* "webp"){
		set $imwebp A;
	}
	if (-f $request_filename.webp) {
		set $imwebp  "${imwebp}B";
	}
	if ($imwebp = AB) {
		rewrite ^(.*) $1.webp;
	}
}
# END Imagify: rewrite rules for webp
```

4- If we access this page https://imagify.rocketlabsqa.ovh/picture-elements/ while display image -> picture element is enabled, we will have fatal error

```
[09-Feb-2024 06:03:54 UTC] PHP Fatal error:  Uncaught Error: Call to undefined method Imagify\Picture\Display::get_cdn_source() in /var/www/imagify.rocketlabsqa.ovh/htdocs/wp-content/plugins/imagify-plugin/classes/Picture/Display.php:725
Stack trace:
#0 /var/www/imagify.rocketlabsqa.ovh/htdocs/wp-content/plugins/imagify-plugin/classes/Picture/Display.php(668): Imagify\Picture\Display->url_to_path()
#1 /var/www/imagify.rocketlabsqa.ovh/htdocs/wp-content/plugins/imagify-plugin/classes/Picture/Display.php(582): Imagify\Picture\Display->get_nextgen_image_data_set()
#2 [internal function]: Imagify\Picture\Display->process_image()
#3 /var/www/imagify.rocketlabsqa.ovh/htdocs/wp-content/plugins/imagify-plugin/classes/Picture/Display.php(431): array_map()
#4 /var/www/imagify.rocketlabsqa.ovh/htdocs/wp-content/plugins/imagify-plugin/classes/Picture/Display.php(130): Imagify\Picture\Display->get_images()
#5 /var/www/imagify.rocketlabsqa.ovh/htdocs/wp-content/plugins/imagify-plugin/classes/Picture/Display.php(105): Imagify\Picture\Display->process_content()
#6 [internal function]: Imagify\Picture\Display->maybe_process_buffer()
#7 /var/www/imagify.rocketlabsqa.ovh/htdocs/wp-includes/functions.php(5373): ob_end_flush()
#8 /var/www/imagify.rocketlabsqa.ovh/htdocs/wp-includes/class-wp-hook.php(324): wp_ob_end_flush_all()
#9 /var/www/imagify.rocketlabsqa.ovh/htdocs/wp-includes/class-wp-hook.php(348): WP_Hook->apply_filters()
#10 /var/www/imagify.rocketlabsqa.ovh/htdocs/wp-includes/plugin.php(517): WP_Hook->do_action()
#11 /var/www/imagify.rocketlabsqa.ovh/htdocs/wp-includes/load.php(1260): do_action()
#12 [internal function]: shutdown_action_hook()
#13 {main}
  thrown in /var/www/imagify.rocketlabsqa.ovh/htdocs/wp-content/plugins/imagify-plugin/classes/Picture/Display.php on line 725
```
5- Using image having Webp or AVIF, then enable display -> picture element. the webp or avif won't be served at FE 
https://imagify.rocketlabsqa.ovh/image-have-webp/ (this is working on trunk)
https://imagify.rocketlabsqa.ovh/image-have-avif/

6- In edit image page if we click generate next-Gen image, we will be redirected to imagify settings page (this isnot happening on trunk)

remyperona committed:
initialize $result variable

remyperona committed:
use filter to get cdn URL

remyperona committed:
use isset

remyperona commented:
Point 1 and 4 should be fixed. For point 3, the rules are as expected, There will be a block for webp and a block for avif depending on what is enabled.

 I can't reproduce point 2.

For point 5 and 6, could we create separate issues for them, as they are not related to the fixes done on this PR.

piotrbak commented:
@Tabrisrp @Mai-Saad For the 3rd point, the [expectation](https://www.notion.so/wpmedia/AVIF-Imagify-Plugin-4a870cab6e0d4232a85962f0b4c1c63a?pvs=4#6e8aa7aaffa540b2bcdceecbb1acec80) here was that we'll be displaying `AVIF > WEBP > Optimized Image` no matter what's enabled. The same should apply for the [Picture method](https://www.notion.so/wpmedia/AVIF-Imagify-Plugin-4a870cab6e0d4232a85962f0b4c1c63a?pvs=4#53f1dfeda1484990a57dbdfbe440f5ab).

This is to handle the scenario when the AVIF is enabled + the browser  is comaptible with WEBP but not with AVIF.
Also when AVIF is enabled, but the AVIF version is not yet generated while we have WEBP present.

Is this possible?

remyperona committed:
update to always add rules when option is enabled

remyperona committed:
update priority

remyperona committed:
correctly update htaccess based on options enabled

remyperona committed:
update condition

remyperona commented:
I updated the code so that the rules are added to the htaccess for both WebP and AVIF, as long as the rewrite option is enabled.

The order in the htaccess is cascading, so the WebP rules are first before the AVIF.

piotrbak commented:
@Tabrisrp @Mai-Saad We should always try to display AVIF as the first one, is this how it'd work with the current order?

Mai-Saad commented:
@Tabrisrp @piotrbak Currently, if we have image with only AVIF version existing on server, using rewrite rules, won't be applied and the AVIF image won't be served on FE. While if we have image with only WebP, the Webp will be served on FE using rewrite rules. While if image have both webP and AVIF, WebP will be served even with picture option.

Note: 
- on the current branch (not on feature/avif), we have duplicate in the rewrite rules text. can you please check this :pray: 
![Screenshot from 2024-02-15 13-31-11](https://github.com/wp-media/imagify-plugin/assets/76941962/69d949c7-66d8-41b1-a4df-472809078ba8)
- For point 5, it was working on feature/avif branch but we can add new GH
- For 6 added new GH as it's same on main feature branch https://github.com/wp-media/imagify-plugin/issues/803

remyperona committed:
invert priorities

remyperona committed:
prevent double display of the NGINX info

remyperona committed:
invert order for webp and avif

remyperona committed:
update sanitize related to convert

remyperona committed:
trigger bulk optimization only when enabling convert_to_avif

remyperona commented:
New update:
- Rewrites rules order updated to correctly prioritize avif, then WebP
- Updated picture display to correctly prioritize avif, then WebP
- Remove double display of conf file notice on NGINX
- Fixed Bulk optimization notice, should only display when enabling `convert_to_avif`

remyperona committed:
Merge branch 'feature/avif' into task/avif-fixes

Mai-Saad commented:
@Tabrisrp Thanks for the update.
- rewrite rules working fine on apache and picture. however, on nginx, if the image has only webP it won't be served (tested on local app using new our generated rewrite rules). Can you please check this and let me know if we need another GH  :pray: 
- regarding bulk notice, it is still displayed on the live site whenever we change settings. Shall we create separate GH for this?
- on nginx, using rewrite rules (on local app) , AVIF type appeared as `octet-stream`, is this expected? (note:rewrite rules on apache will show type as `avif`)
![Screenshot from 2024-02-16 13-29-08](https://github.com/wp-media/imagify-plugin/assets/76941962/52be7a12-3024-477a-99e2-c959b43bf00b)

remyperona commented:
> however, on nginx, if the image has only webP it won't be served (tested on local app using new our generated rewrite rules). Can you please check this and let me know if we need another GH

I don't know why it would not work, it's the same rules as before. I have no knowledge on NGINX configuration, I'm not sure how to help on that.

> regarding bulk notice, it is still displayed on the live site whenever we change settings. Shall we create separate GH for this?

I can't reproduce it

> on nginx, using rewrite rules (on local app) , AVIF type appeared as octet-stream, is this expected? (note:rewrite rules on apache will show type as avif)

I found the following in ShortPixel documentation:

You basically need to look for the file named mime.types, which is usually located in the root of the NGINX installation (/etc/nginx/), and add the two MIME types: image/avif and image/avif-sequence

piotrbak commented:
```
however, on nginx, if the image has only webP it won't be served (tested on local app using new our generated rewrite rules). Can you please check this and let me know if we need another GH
```
@Mai-Saad Does it work on the trunk?

```
regarding bulk notice, it is still displayed on the live site whenever we change settings. Shall we create separate GH for this?

```
@Mai-Saad Did you try clearing all the transients? Does it happen on the Rocketlabs or local environment (or both)?

Mai-Saad commented:
> https://github.com/wp-media/wp-rocket/issues/6312

working fine on trunk using local env

> @Mai-Saad Did you try clearing all the transients? Does it happen on the Rocketlabs or local environment (or both)?

It is happening on https://imagify.rocketlabsqa.ovh/ but not local (same after delete transients)

piotrbak commented:
@Tabrisrp Could you take a look at this environment?

@Mai-Saad @Tabrisrp We'll tacke the NGINX rewrite rules [here](https://github.com/orgs/wp-media/projects/109/views/1?pane=issue&itemId=53703274).

Mai-Saad commented:
> You basically need to look for the file named mime.types, which is usually located in the root of the NGINX installation (/etc/nginx/), and add the two MIME types: image/avif and image/avif-sequence

@Tabrisrp adding `    image/avif                               avif;` in `conf/nginx/mime-types.conf.hbs` works while using picture and rewrite on local env :tada: :pray:
